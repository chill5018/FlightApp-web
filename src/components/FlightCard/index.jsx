import React from 'react';
import PropTypes from 'prop-types';
import { Button } from 'reactstrap';
import { Link } from 'react-router-dom';


const propTypes = {
  flight: PropTypes.shape({
    departureFlight: PropTypes.shape({
      airline: PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
      }).isRequired,
      arrivalDateTime: PropTypes.string.isRequired,
      createdAt: PropTypes.string.isRequired,
      departureDateTime: PropTypes.string.isRequired,
      destination: PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        code: PropTypes.string.isRequired,
      }).isRequired,
      id: PropTypes.number.isRequired,
      flightNumber: PropTypes.string.isRequired,
      origin: PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        code: PropTypes.string.isRequired,
      }).isRequired,
      updatedAt: PropTypes.string.isRequired,
    }).isRequired,
    returnFlight: PropTypes.shape({
      airline: PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
      }).isRequired,
      arrivalDateTime: PropTypes.string.isRequired,
      createdAt: PropTypes.string.isRequired,
      departureDateTime: PropTypes.string.isRequired,
      destination: PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        code: PropTypes.string.isRequired,
      }).isRequired,
      id: PropTypes.number.isRequired,
      flightNumber: PropTypes.string.isRequired,
      origin: PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        code: PropTypes.string.isRequired,
      }).isRequired,
      updatedAt: PropTypes.string.isRequired,
    }),
  }),
};

const defaultProps = {
  flight: {
    returnFlight: undefined,
  },
};

const FlightCard = ({ flight }) => (
  <section className="app-main-flightCard">
    <div className="app-main-col app-main-flightCard-icon app-main-flightCard-icon-plane" />
      <h6 className="app-main-col app-main-flightCard-icon-title">
        { flight.departureFlight.airline.name }
      </h6>
    <div className="app-main-col app-main-flightCard-icon app-main-flightCard-icon-ticket" />
      <h6 className="app-main-col app-main-flightCard-icon-title">
        Economy
      </h6>
    <div className="app-main-flightCard-hr" />

    <div className="app-main-flightCard-segment">
    <div className="app-main-col app-main-col_1 app-main-flightCard-legRow">
      <div className="app-main-col">
        <h5 className="app-main-col app-main-flightCard-airportCode">
          { flight.departureFlight.origin.code.toUpperCase() }
        </h5>
        <h5 className="app-main-col app-main-flightCard-time">
          18:30
        </h5>
        <br />
        <h6 className="app-main-col app-main-flightCard-airport">
          { flight.departureFlight.origin.name }
        </h6>
      </div>

      <div className="app-main-col">
        <p className="app-main-flightCard-leg-time">
          8h 30m
        </p>
        <div className="app-main-flightCard-leg" />
        <p className="app-main-flightCard-leg-type">
          Direct
        </p>
      </div>

      <div className="app-main-col">
        <h5 className="app-main-col app-main-flightCard-time">
          21:00
        </h5>
        {' '}
        <h5 className="app-main-col app-main-flightCard-airportCode">
          { flight.departureFlight.destination.code.toUpperCase() }
        </h5>
        <br />
        <h6 className="app-main-col app-main-flightCard-airport">
          { flight.departureFlight.destination.name }
        </h6>
      </div>
    </div>

    { flight.returnFlight && (
      <div className="app-main-col app-main-col_1 app-main-flightCard-legRow">
        <div className="app-main-col">
          <h5 className="app-main-col app-main-flightCard-airportCode">
            { flight.returnFlight.origin.code.toUpperCase() }
          </h5>
          <h5 className="app-main-col app-main-flightCard-time">
            23:00
          </h5>
          <br />
          <h6 className="app-main-col app-main-flightCard-airport">
          { flight.returnFlight.origin.name }
          </h6>
        </div>

        <div className="app-main-col">
          <p className="app-main-flightCard-leg-time">
            7h 10m
          </p>
          <div className="app-main-flightCard-leg" />
          <p className="app-main-flightCard-leg-type">
          Direct
          </p>
        </div>

        <div className="app-main-col">
          <h5 className="app-main-col app-main-flightCard-time">
            12:10
          </h5>
          {' '}
          <h5 className="app-main-col app-main-flightCard-airportCode">
          { flight.returnFlight.destination.code.toUpperCase() }
          </h5>
          <br />
          <h6 className="app-main-col app-main-flightCard-airport">
          { flight.returnFlight.destination.name }
          </h6>
        </div>
      </div>
    )}

    </div>

    <div className="app-main-flightCard-pass">
      <h3 className="app-main-flightCard-pass-price">
        2.236 DKK
      </h3>
      <p className="app-main-flightCard-pass-p">
        Flight
      </p>
      <Button className="w-100 mt-5 app-main-flightCard-pass-btn">
        Book
      </Button>
    </div>

  </section>
);

export default FlightCard;

FlightCard.propTypes = propTypes;
FlightCard.defaultProps = defaultProps;
