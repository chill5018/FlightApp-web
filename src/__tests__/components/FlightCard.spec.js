import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import FlightCard from '../../components/FlightCard';

const flight = {
  price: 1234.12,
  departureFlight:
  {
    id: 12,
    flightNumber: 'SAS-1006',
    departureDateTime: '2019-02-02T04:00:00.012Z',
    arrivalDateTime: '2019-02-02T16:00:00.012Z',
    createdAt: '2018-12-12T21:17:59.617Z',
    updatedAt: '2018-12-12T21:17:59.617Z',
    airline: {
      id: 3,
      name: 'SAS',
    },
    destination: {
      id: 5,
      name: 'New York City John F Kennedy',
      code: 'JFK',
    },
    origin: {
      id: 1,
      name: 'Copenhagen Airport',
      code: 'CPH',
    },
  },
  returnFlight: {
    id: 30,
    flightNumber: 'SAS-1024',
    departureDateTime: '2019-02-14T04:00:00.012Z',
    arrivalDateTime: '2019-02-14T16:00:00.012Z',
    createdAt: '2018-12-12T21:21:02.879Z',
    updatedAt: '2018-12-12T21:21:02.879Z',
    airline: {
      id: 3,
      name: 'SAS',
    },
    destination: {
      id: 1,
      name: 'Copenhagen Airport',
      code: 'CPH',
    },
    origin: {
      id: 5,
      name: 'New York City John F Kennedy',
      code: 'JFK',
    },
  },
};

describe('Flight Card Snapshot', () => {
  it('renders correctly', () => {
    const component = renderer.create(
      <FlightCard key={flight.id} flight={flight} />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Flight Card', () => {
  const component = shallow(<FlightCard flight={flight} />);

  describe('Departure Flight', () => {
    it('Renders Correct Departure Airport Code', () => {
      const expectedValue = 'CPH';
      expect(component.find('.app-main-flightCard-airportCode').at(0).text()).toEqual(expectedValue);
    });

    it('Renders Correct Arrival Airport Code', () => {
      const expectedValue = 'JFK';
      expect(component.find('.app-main-flightCard-airportCode').at(1).text()).toEqual(expectedValue);
    });

    it('Renders Correct Departure Time', () => {
      const expectedValue = '04:00';
      expect(component.find('.app-main-flightCard-time').at(0).text()).toEqual(expectedValue);
    });

    it('Renders Correct Duration', () => {
      const expectedValue = '12h 0m';
      expect(component.find('.app-main-flightCard-leg-time').at(0).text()).toEqual(expectedValue);
    });
  });

  describe('Return Flight', () => {
    it('Renders Correct Departure Airport Code', () => {
      const expectedValue = 'JFK';
      expect(component.find('.app-main-flightCard-airportCode').at(2).text()).toEqual(expectedValue);
    });

    it('Renders Correct Arrival Airport Code', () => {
      const expectedValue = 'CPH';
      expect(component.find('.app-main-flightCard-airportCode').at(3).text()).toEqual(expectedValue);
    });

    it('Renders Correct Departure Time', () => {
      const expectedValue = '16:00';
      expect(component.find('.app-main-flightCard-time').at(1).text()).toEqual(expectedValue);
    });

    it('Renders Correct Duration', () => {
      const expectedValue = '12h 0m';
      expect(component.find('.app-main-flightCard-leg-time').at(1).text()).toEqual(expectedValue);
    });
  });

  it('Renders Correct Price', () => {
    const expectedValue = '1234.12 DKK';
    expect(component.find('.app-main-flightCard-pass-price').text()).toEqual(expectedValue);
  });

  it('Generates Price', () => {
    const newFlight = flight;
    delete newFlight.price;

    const priceComponent = shallow(<FlightCard flight={newFlight} />);
    const result = priceComponent.find('.app-main-flightCard-pass-price').text();
    const price = +result.split(' ')[0];

    expect(price).toBeGreaterThan(0);
  });
});
