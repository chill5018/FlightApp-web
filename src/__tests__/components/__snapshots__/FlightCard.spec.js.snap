// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flight Card Snapshot renders correctly 1`] = `
<section
  className="app-main-flightCard"
>
  <div
    className="app-main-col app-main-flightCard-icon app-main-flightCard-icon-plane"
  />
  <h6
    className="app-main-col app-main-flightCard-icon-title"
  >
    SAS
  </h6>
  <div
    className="app-main-col app-main-flightCard-icon app-main-flightCard-icon-ticket"
  />
  <h6
    className="app-main-col app-main-flightCard-icon-title"
  >
    Economy
  </h6>
  <div
    className="app-main-flightCard-hr"
  />
  <div
    className="app-main-flightCard-segment"
  >
    <div
      className="app-main-col app-main-col_1 app-main-flightCard-legRow"
    >
      <div
        className="app-main-col"
      >
        <h5
          className="app-main-col app-main-flightCard-airportCode"
        >
          CPH
        </h5>
        <h5
          className="app-main-col app-main-flightCard-time"
        >
          04:00
        </h5>
        <br />
        <h6
          className="app-main-col app-main-flightCard-airport"
        >
          Copenhagen Airport
        </h6>
      </div>
      <div
        className="app-main-col"
      >
        <p
          className="app-main-flightCard-leg-time"
        >
          12h 0m
        </p>
        <div
          className="app-main-flightCard-leg"
        />
        <p
          className="app-main-flightCard-leg-type"
        >
          Direct
        </p>
      </div>
      <div
        className="app-main-col"
      >
        <h5
          className="app-main-col app-main-flightCard-time"
        >
          16:00
        </h5>
         
        <h5
          className="app-main-col app-main-flightCard-airportCode"
        >
          JFK
        </h5>
        <br />
        <h6
          className="app-main-col app-main-flightCard-airport"
        >
          New York City John F Kennedy
        </h6>
      </div>
    </div>
    <div
      className="app-main-col app-main-col_1 app-main-flightCard-legRow"
    >
      <div
        className="app-main-col"
      >
        <h5
          className="app-main-col app-main-flightCard-airportCode"
        >
          JFK
        </h5>
        <h5
          className="app-main-col app-main-flightCard-time"
        >
          04:00
        </h5>
        <br />
        <h6
          className="app-main-col app-main-flightCard-airport"
        >
          New York City John F Kennedy
        </h6>
      </div>
      <div
        className="app-main-col"
      >
        <p
          className="app-main-flightCard-leg-time"
        >
          12h 0m
        </p>
        <div
          className="app-main-flightCard-leg"
        />
        <p
          className="app-main-flightCard-leg-type"
        >
          Direct
        </p>
      </div>
      <div
        className="app-main-col"
      >
        <h5
          className="app-main-col app-main-flightCard-time"
        >
          16:00
        </h5>
         
        <h5
          className="app-main-col app-main-flightCard-airportCode"
        >
          CPH
        </h5>
        <br />
        <h6
          className="app-main-col app-main-flightCard-airport"
        >
          Copenhagen Airport
        </h6>
      </div>
    </div>
  </div>
  <div
    className="app-main-flightCard-pass"
  >
    <h3
      className="app-main-flightCard-pass-price"
    >
      1234.12 DKK
    </h3>
    <p
      className="app-main-flightCard-pass-p"
    >
      Flight
    </p>
    <button
      aria-label={null}
      className="w-100 mt-5 app-main-flightCard-pass-btn btn btn-secondary"
      onClick={[Function]}
    >
      Book
    </button>
  </div>
</section>
`;
